### DISCUSSION TITLE
Usage of continuous in GitOps Principle Nr. 3

### DISCUSSION THREAD

**Discussion by murillodigital** [open-gitops/project#80] [General]
The following living document reflects the current state of definition around the core principles of the GitOps model [https://hackmd.io/arwvV8NUQX683uBM3HzyNQ](https://hackmd.io/arwvV8NUQX683uBM3HzyNQ), as collaboratively defined by the community and the GitOps Principles Committee.

This discussion pertains Principle 3 and the use of the word _continuous_ as a component of the title.

An argument has been raised as to the inaccuracy or impossibility of **continuous** being a determining factor of the GitOps reconciliation process, considering concerns such as incidents, freeze windows as well as other corner cases in which stating a fixed expectation of continuity may not reflect the true nature of production and enterprise systems.

Discussions surrounding this subject can be found in:

Mar 29, 2021
-  [Google Doc](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.lfuaf62iclbk)
- [YouTube](https://www.youtube.com/watch?v=2VpbWrKjDkQ)
- [Zoom](https://zoom.us/rec/play/xyvAI-m4L4gpSjnDakhXA6uie6A6ueEDyAd3zl-pMx8j3a--JCfIHE4ZjCAfK_sgJr8FDqKE0_Jy7A.RE6k3oSUL9HY2yfg)

April 05, 2021
- [Google Doc](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.mgg85m8p2zrf)
- [YouTube](https://www.youtube.com/watch?v=SGcSRTWnT3k)
- [Zoom](https://zoom.us/rec/share/kKmbwV-AaMnqv4f035BrVgvzXXpB0nks9k74tgDNEOM5MgwwkVDgcY_B8xyAF-M6.1ljFW-gp9-TUJwTZ)

April 14, 2021
- [Google Doc](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.7wmerton06if)
- [YouTube](https://www.youtube.com/watch?v=zWxAZuHNGYM)

Please use this discussion to communicate your position as succinctly as possible around the use of the word continuous in the principle, as well as the implications considering your own experience with as well as expectations around GitOps.

**Comment by moshloop** [comment]
My objection to the use of continuous is around the dictionary definition of `uninterrupted and without gaps`

1) Continuous is an impossible state for a distributed system,  even the fastest reconciliation loops will happen in a few seconds and while not a large gap or interruption it is a gap nonetheless

    This distinction has already come back to bite the continuous delivery movement with a split into continuous deployment and continuous delivery (with continuously deliverable being the now accepted meaning)
  
    Even continuous integration falls into this trap (CI doesn't mean we are continuously building or running a daily build, it means the state of development is continuously integrateble (and code sitting on a developers workstation not being checked in, is not integrateble) 

   A principle should be unassailable and have no room for interpretation - refuting the dictionary definition of a term, defeats the purpose of a principle - something that succinctly explains a concept or idea 

2) Reconciliation loops MUST run continuously to check the state, but the reconciliation action cannot physically and should not run continuously, a [PID](https://en.wikipedia.org/wiki/PID_controller#:~:text=A%20proportional%E2%80%93integral%E2%80%93derivative%20controller,continuously%20calculates%20an%20error%20value) based reconciliation loop (which is what you need to not shoot yourself in the foot) by design inserts gaps and interruptions based on the feedback from the system

3) Even if the continuous application of state is desirable (which control theory states it is not) it is not pragmatic, from my understanding the WG was brought together to increase the adoption levels of GitOps, you don't do that by raising the barrier to entry to a point that nothing in reality actually meets the criteria set forward. 

We should take a step back and ask what are the current barriers to entry and how do we remove them,  not adding new barriers.

**Reply by theWilliamChia** [reply]
@moshloop Everything you are saying makes complete sense to me. I'm not sure why GitOps couldn't simply have a principle of "state reconciliation" rather than "continuous state reconciliation". 

This stood out to me 

> WG was brought together to increase the adoption levels of GitOps, you don't do that by raising the barrier to entry to a point that nothing in reality actually meets the criteria set forward.

On this topic, I'm curious, what if reconciliation was periodic, or even regular, rather than continuous? These seem like states of maturity along a GitOps continuum rather than a binary yes/no. 

It doesn't seem helpful to tell someone, "No you're not doing GitOps, you need to be at this level of maturity before you are really doing GitOps." 

It seems more helpful to say, "these are the core principles and if you have the kernel of those, that's GitOps, and oh by the way, we can help move you to the next level." That feels similar to how DevOps is.

**Reply by schlomo** [reply]
@moshloop thanks a lot for writing down your thoughts, now I understand much better the point that bothers you.

Do we agree on the fact that the checking for deviation should run continuously, your pint (2)? That was my motivation to wish for "continuous" to be part of the principle.

If I understand your arguments correctly, It seems to me that a wording like "The principle of continuous state reconciliatiatability" (or however in correct English) would be more appropriate in order to reflect - analog to CD meaning continuously deliverable - that it is in principle possible to reconcile the state at any given moment.

So as a compromise, could we somehow rephrase the principle to make it clear that the checking happens all the time - continuously - and that then the decision to reconcile includes also the option to not reconcile at that moment?

> Software agents continuously, and automatically, compare a system’s Actual State to its Desired State. If the actual and desired states differ for any reason, automated actions to reconcile them are initiated.

The principle now has two sentences. The first one mentions continuous checking and the second implies immediate action. Would changing the second sentence to something like this better express what you meant?

If the actual and desired states differ for any reason, automated actions to reconcile them **can be** initiated.

Or is that already too weak language? What would you suggest?

About the heading of the principle: I think that if "continuously" stays in the first sentence then it can be omitted in the heading as the continuous checking is the important part in my opinion.

**Comment by moshloop** [comment]
> On this topic, I'm curious, what if reconciliation was periodic, or even regular, rather than continuous? These seem like states of maturity along a GitOps continuum rather than a binary yes/no.

I would argue that most if not all reconciliation are already periodic and regular (e.g flux polls for commits every 5m by default) - Webhooks can speed up the loop to poll sooner

> It doesn't seem helpful to tell someone, "No you're not doing GitOps, you need to be at this level of maturity before you are really doing GitOps."

Completely agree, we need to define a baseline e.g. Not reliant on webhooks or manually triggered jobs and then offer recommendations that increase maturity (but not on a mature scale)

> Do we agree on the fact that the checking for deviation should run continuously, your pint (2)? That was my motivation to wish for "continuous" to be part of the principle.

100% 

>  The principle of continuous state reconciliatiatability

I am not sure the *ability maps into GitOps, The state should always be reconciliable (through the use of linting and unit tests) but in addition, the checking should also run in a loop

> Software agents continuously, and automatically, compare a system’s Actual State to its Desired State. If the actual and desired states differ for any reason, automated actions to reconcile them are initiated.

This description is ok as *initiated* doesn't imply completion - @todaywasawesome suggested *automated actions attempt to reconcile them*  which I like because it implies that an attempt could fail e.g. for incidents, change freezes, etc..


What I would suggest is 

**3. The principle of state reconciliation loops**
Software agents continuously, and automatically, compare a system’s Actual State to its Desired State. If the actual and desired states differ for any reason, automated actions attempt to reconcile them

- The desired state **MUST** be continuously reconciliable through the use of linting/unit tests
- The desired state **MUST** be continuously polled for changes (which can be accelerated through webhooks, but webhooks alone is not GitOps)
- The desired and actual state **SHOULD** be compared continuously to detect drift and feedback into the state store.
- The actual state **SHOULD** feedback into the state store (e.g. Github Deployment tracking the rollout status)
- The reconciliation loop **SHOULD** limit logic around policy (e.g. for change freeze periods should be implemented in merge blocking tests on pull requests)

**Reply by lotharschulz** [reply]
I wonder if a crisp principle title like **The principle of state reconciliation** is enough.

@moshloop's proposal principle description

> Software agents continuously, and automatically, compare a system’s Actual State to its Desired State. If the actual and desired states differ for any reason, automated action attempts to reconcile them.

can be implemented with loops though. However other options might be possible as well.

**Reply by schlomo** [reply]
> **3. The principle of state reconciliation loops**
Software agents continuously, and automatically, compare a system’s Actual State to its Desired State. If the actual and desired states differ for any reason, automated actions attempt to reconcile them

I am happy to agree with this version.

**Reply by lloydchang** [reply]
@moshloop wrote:
> The reconciliation loop **SHOULD** limit logic around policy (e.g. for change freeze periods should be implemented in merge blocking tests on pull requests)
-----
@moshloop ❓ Is the intent to discourage logic around policy such as the following? Thanks.
> If there is an active matching `allow` and an active matching `deny` then syncs will be denied as `deny` windows override `allow` windows.
> https://argo-cd.readthedocs.io/en/stable/user-guide/sync_windows/

**Reply by moshloop** [reply]
I am personally indifferent to whether this should be discouraged or not (as long as it is still allowed) others have a view that this should be done as part of the pull request workflow and the commits should only merge inside the change/sync period

**Reply by lloydchang** [reply]
LGTM

**Comment by moshloop** [comment]
> can be implemented with loops though. However other options might be possible as well.

Could you elaborate on this? I am struggling to think of any other option?

**Reply by lotharschulz** [reply]
Watchers observe _desired_ and _actual_ state. Reconciliation action attempts are triggered if states' drift is observed. 

---

In the call March 29th 2021, I understood from @bricef that GitOps principles do not contain references to _event systems_ by purpose. 

[hack md document#Principle-3-Notes](https://hackmd.io/arwvV8NUQX683uBM3HzyNQ#Principle-3-Notes) reads:

> “continuous?” - subtleties with closed/open loops in system and control theory, what about event driven stuff?

First time I found the quote in the [hack md document](https://hackmd.io/arwvV8NUQX683uBM3HzyNQ) history is _Mon, Apr 5, 2021 9:20 PM_

**Reply by moshloop** [reply]
I think it is the other options like event-based systems that we want to exclude, and adding in the term _loop_ achieves that

**Reply by lotharschulz** [reply]
I see your point and I agree with our discussion as context. Without our discussion as context I had a hard time to derive a conclusion like "adding in the term loop excludes event-based systems". What if we explicitly describe that event-based systems are excluded? Such a description could be part of a _principles notes_ section.

**Reply by lloydchang** [reply]
@bricef @moshloop @lotharschulz 

Are you suggesting to exclude the following?
> • The Notification Controller is a Kubernetes operator, specialized in handling inbound and outbound events.
> • The controller handles events coming from external systems (GitHub, GitLab, Bitbucket, Harbor, Jenkins, etc) and notifies the GitOps toolkit controllers about source changes.
> • The controller handles events emitted by the GitOps toolkit controllers (source, kustomize, helm) and dispatches them to external systems (Slack, Microsoft Teams, Discord, Rocker) based on event severity and involved objects.
> https://toolkit.fluxcd.io/components/notification/controller/

**Reply by moshloop** [reply]
I think the outbound events (unless they are back to the git repository) are supplementary and neither included nor excluded. 

If the notification controller is the only trigger for state reconciliation, then yes it would be excluded, it can be in addition to a polling reconciliation, but not instead of.

> What if we explicitly describe that event-based systems are excluded? Such a description could be part of a principles notes section.

I agree we should elaborate more in the description,

**Reply by lloydchang** [reply]
@moshloop wrote:
> If the notification controller is the only trigger for state reconciliation, then yes it would be excluded, it can be in addition to a polling reconciliation, but not instead of.

My understanding is the notification controller's webhook receivers are additions, based on the following written by @stefanprodan:
> [**Webhook receivers**](https://github.com/fluxcd/notification-controller/blob/c058cce0ea4c96b76e3649c3a3fc40b1be7c53b4/docs/spec/README.md#webhook-receivers)
> GitOps controllers are by nature pull based, in order to notify the controllers about changes in Git or Helm repositories, one may wish to setup webhooks and trigger a cluster reconciliation every time a source changes.

> [**Setup Webhook Receivers**](https://github.com/fluxcd/flux2/blob/97592a138785d67b27641dc628169016bc447ad4/docs/guides/webhook-receivers.md#setup-webhook-receivers)
The GitOps toolkit controllers are by design **pull-based**. In order to notify the controllers about changes in Git or Helm repositories, you can setup webhooks and trigger a cluster reconciliation every time a source changes. Using webhook receivers, you can build **push-based** GitOps pipelines that react to external events.

Thank you @moshloop @stefanprodan @lotharschulz @bricef

**Comment by lloydchang** [comment]
@murillodigital wrote:
> Discussions surrounding this subject can be found in the following recordings: [here](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.mgg85m8p2zrf) and here (link coming soon).

@murillodigital When your time permits, please amend the above as follows:
> Discussions are in:
>
> Mar 29, 2021
> • [Google Doc](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.lfuaf62iclbk)
> • [YouTube](https://www.youtube.com/watch?v=2VpbWrKjDkQ)
> • [Zoom](https://zoom.us/rec/play/xyvAI-m4L4gpSjnDakhXA6uie6A6ueEDyAd3zl-pMx8j3a--JCfIHE4ZjCAfK_sgJr8FDqKE0_Jy7A.RE6k3oSUL9HY2yfg)
>
> April 05, 2021
> • [Google Doc](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.mgg85m8p2zrf)
> • [YouTube](https://www.youtube.com/watch?v=SGcSRTWnT3k)
> • [Zoom](https://zoom.us/rec/share/kKmbwV-AaMnqv4f035BrVgvzXXpB0nks9k74tgDNEOM5MgwwkVDgcY_B8xyAF-M6.1ljFW-gp9-TUJwTZ)

> April 14, 2021
> • [Google Doc](https://docs.google.com/document/d/1hxifmCdOV5_FbKloDJRWZQHq0ge-trXJKF-BgV4wHVk/edit#heading=h.7wmerton06if)
> • [YouTube](https://www.youtube.com/watch?v=zWxAZuHNGYM)

Thank you!

**Reply by scottrigby** [reply]
Added, thanks for the reminder

**Reply by lloydchang** [reply]
Thanks @theWilliamChia for [clarifying](https://cloud-native.slack.com/archives/C01G9DEE88M/p1618513877045300) that the link was outdated.

Thanks @scottrigby for adding additional links.

**Comment by scottrigby** [comment]
For posterity, the Principles Committee has since agreed to keep "continuous" as long as we keep a prominent note about this colocated with the principles. As of v0.1.0 of the OpenGitOps documents: <https://github.com/open-gitops/documents/blob/v0.1.0/PRINCIPLES.md> this was addressed (see [Continuous glossary item](https://github.com/open-gitops/documents/blob/v0.1.0/PRINCIPLES.md#continuous)).
